File: frameworks/Java/fair/java-spring-jdbc-template/src/main/java/io/techempower/benchmark/spring/repository/WorldRepository.java
package io.techempower.benchmark.spring.repository;

import io.techempower.benchmark.spring.model.Fortune;
import io.techempower.benchmark.spring.model.World;
import org.springframework.jdbc.core.JdbcTemplate;
import org.springframework.jdbc.core.RowMapper;
import org.springframework.stereotype.Repository;

import java.util.List;

@Repository
public class WorldRepository {

    private final JdbcTemplate jdbcTemplate;

    private static final RowMapper<World> WORLD_ROW_MAPPER = (rs, rowNum) ->
            new World(rs.getInt("id"), rs.getInt("randomnumber"));

    private static final RowMapper<Fortune> FORTUNE_ROW_MAPPER = (rs, rowNum) ->
            new Fortune(rs.getInt("id"), rs.getString("message"));

    public WorldRepository(JdbcTemplate jdbcTemplate) {
        this.jdbcTemplate = jdbcTemplate;
    }

    public World findById(int id) {
        return jdbcTemplate.queryForObject(
                "SELECT id, randomnumber FROM world WHERE id = ?",
                WORLD_ROW_MAPPER,
                id
        );
    }

    public int findRandomNumberById(int id) {
        return jdbcTemplate.queryForObject(
                "SELECT randomnumber FROM world WHERE id = ?",
                Integer.class,
                id
        );
    }

    public void update(World world) {
        jdbcTemplate.update(
                "UPDATE world SET randomnumber = ? WHERE id = ?",
                world.randomNumber(),
                world.id()
        );
    }

    public List<Fortune> findAllFortunes() {
        return jdbcTemplate.query(
                "SELECT id, message FROM fortune",
                FORTUNE_ROW_MAPPER
        );
    }
}

File: frameworks/Java/fair/kotlin-ktor-cio-jdbc/src/main/kotlin/io/techempower/benchmark/ktor/repository/WorldRepository.kt
package io.techempower.benchmark.ktor.repository

import com.zaxxer.hikari.HikariConfig
import com.zaxxer.hikari.HikariDataSource
import io.techempower.benchmark.ktor.model.Fortune
import io.techempower.benchmark.ktor.model.World
import io.techempower.benchmark.ktor.util.QueryUtils
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import java.sql.Connection

class WorldRepository {
    private val dataSource = HikariDataSource(HikariConfig().apply {
        jdbcUrl = env("POSTGRES_JDBC_URL", "jdbc:postgresql://localhost:5432/postgres")
        username = env("POSTGRES_USER", "postgres")
        password = env("POSTGRES_PASS", "postgres")
        driverClassName = "org.postgresql.Driver"
        poolName = "ktor-jdbc"
        minimumIdle = 16
        maximumPoolSize = 512
        connectionTimeout = 10_000
        validationTimeout = 5_000
        idleTimeout = 0
        maxLifetime = 0
        initializationFailTimeout = 10_000
        addDataSourceProperty("preparedStatementCacheQueries", "512")
        addDataSourceProperty("preparedStatementCacheSizeMiB", "16")
        addDataSourceProperty("prepareThreshold", "1")
        addDataSourceProperty("loggerLevel", "OFF")
        addDataSourceProperty("sslmode", "disable")
        addDataSourceProperty("disableColumnSanitiser", "true")
    })

    suspend fun findWorld(id: Int): World = withContext(Dispatchers.IO) {
        connection().use { connection ->
            findWorld(connection, id)
        }
    }

    suspend fun findWorlds(count: Int): List<World> = withContext(Dispatchers.IO) {
        connection().use { connection ->
            val worlds = ArrayList<World>(count)
            repeat(count) {
                worlds.add(findWorld(connection, QueryUtils.randomWorld()))
            }
            worlds
        }
    }

    suspend fun updateWorlds(count: Int): List<World> = withContext(Dispatchers.IO) {
        connection().use { connection ->
            connection.autoCommit = false
            try {
                val worlds = ArrayList<World>(count)
                repeat(count) {
                    val oldWorld = findWorld(connection, QueryUtils.randomWorld())
                    worlds.add(World(oldWorld.id, QueryUtils.randomWorld(oldWorld.randomNumber)))
                }
                worlds.sortBy(World::id)
                updateWorlds(connection, worlds)
                connection.commit()
                worlds
            } catch (e: Exception) {
                connection.rollback()
                throw e
            }
        }
    }

    suspend fun findFortunes(): List<Fortune> = withContext(Dispatchers.IO) {
        connection().use { connection ->
            connection.prepareStatement("SELECT id, message FROM fortune").use { statement ->
                statement.executeQuery().use { resultSet ->
                    val fortunes = ArrayList<Fortune>()
                    while (resultSet.next()) {
                        fortunes.add(Fortune(resultSet.getInt(1), resultSet.getString(2)))
                    }
                    fortunes
                }
            }
        }
    }

    private fun connection(): Connection = dataSource.connection

    private fun findWorld(connection: Connection, id: Int): World {
        connection.prepareStatement("SELECT id, randomnumber FROM world WHERE id = ?").use { statement ->
            statement.setInt(1, id)
            statement.executeQuery().use { resultSet ->
                resultSet.next()
                return World(resultSet.getInt(1), resultSet.getInt(2))
            }
        }
    }

    private fun updateWorlds(connection: Connection, worlds: List<World>) {
        connection.prepareStatement("UPDATE world SET randomnumber = ? WHERE id = ?").use { statement ->
            for (world in worlds) {
                statement.setInt(1, world.randomNumber)
                statement.setInt(2, world.id)
                statement.addBatch()
            }
            statement.executeBatch()
        }
    }

    private fun env(name: String, fallback: String): String {
        val value = System.getenv(name)
        return if (value.isNullOrBlank()) fallback else value
    }
}
